Errors - warnings - context

matthid opened the pull request:
This PR adds 

- some context information to the AST. This is generic because it will start with a `SyntaxNode` (this PR) and can be mapped to range information (or whatever we need later).
- Collects "results" in the processing steps to collect errors, warnings (and later hints)

I probably didn't manage to select the "correct" `SyntaxNode` on every AST-item but it should be good enough to start.
I'm not sure if this is the best approach, on some occasions I had to refactor the code because F# could not inference the types properly (`let f =`).

matthid commented:
> I am perfectly fine with collecting errors.

Yes what I mean is the CLI should report them properly (currently just `%A`) and return non-zero exit code...

matthid commented:
Oh and the error/warning collection probably works at the moment but I probably completely failed to collecting them in the correct order...

matthid commented on cs2fs/cs2fs.fs:
```diff
@@ -416,15 +416,15 @@ let rec convertNode tryImplicitConv (model: SemanticModel) (node: SyntaxNode) :
             | "++" -> res, withOp "+" "1"
             | "--" -> res, withOp "-" "1"
-            | x -> ErrorType.UnknownPostfixOperator x |> createError node |> addError res, Mk.mkError node
+            | x -> ErrorType.UnknownPostfixOperator x |> createError node |> addError res, Mk.mkError node (sprintf "UnknownPostfixOperator %s" x)
```
I actually kind of left that out on purpose, basically you should always have a corresponding error with details and you can always correlate back (because of the node) if needed?

But I have no problem with the message, just feels duplicated. I guess you plan to use that somehow in the output generation? Then I completely agree that we need something here, but It could be a 'DU' as well?

jindraivanek replied:
I wanted some way to have error details in output instead of only (* ERROR *), and went by easiest route. Maybe better will be to add `AdditionalInfo : string` to `ConvertError` and then find error by node?

matthid replied:
From a design perspective I'm not sure what the best approach is and if the user has any benefit in `(* XZY failed/not supported *)` (considering he will get a list of errors anyway). Maybe we could even just print the C# source here.

Errors will definitely happen that's why I think it's reasonably to start with that.
Maybe @JohBa has some ideas on how we can visualize the error case (browser & cli)?

matthid replied:
> Maybe better will be to add AdditionalInfo : string to ConvertError and then find error by node?

Another aspect is that I'm not sure if parsing is the correct place to decide for an definite output. I'd assume that in the future we want to represent stuff differently in the browser than in the CLI (This is the reason why I added the DU for the error case instead of an error message)

jindraivanek replied:
There is cases that cant be easily rewritten to F#, for example `break`, `continue`, early `return`. From CLI POV it is important that user know this is the place where `break` was in C# code.

matthid replied:
@jindraivanek I've planned to address at least some of them, but I agree.

jindraivanek replied:
I will rework this to use ConvertResult and remove message from ExprError.

matthid commented:
Found some issues and fixed them. Note: I can only recommend doing testing by starting the server and using the browser. It's really fast. I can document the setup if you are interested (I added the configuration for visual studio code to the other repository).

matthid commented:
> but we might need some tests soonish, but I feel like it is too early for text/ast comparisons as the stuff is still moving...

Instead of string comparison we could compile and run the C# and the F# and compare the result (output). Until we have something more stable that should save us from regression. I'll try to setup something. Should we add a CI to this repository as well?

jindraivanek commented:
>  run the C# and the F# and compare the result (output)

This idea was the reason why I added Rosetta code C# samples to repo :)

> Should we add a CI to this repository as well?

I would only welcome it.
